test/cri-containerd/removepodsandbox_test.go

```diff
@@ -0,0 +1,156 @@
+// +build functional
+
+package cri_containerd
+
+import (
+	"context"
+	"testing"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	runtime "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+)
+
+type TestConfig struct {
+	name             string
+	containerName    string
+	requiredFeatures []string
+	runtimeHandler   string
+	sandboxImage     string
+	containerImage   string
+	cmd              []string
+}
+
+// Utility function to test removal of a sandbox with no containers and no previous call to stop the pod
+func runPodSandboxTestWithoutPodStop(t *testing.T, request *runtime.RunPodSandboxRequest) {
+	client := newTestRuntimeClient(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	podId := runPodSandbox(t, client, ctx, request)
+	removePodSandbox(t, client, ctx, podId)
+
+	_, err := client.PodSandboxStatus(ctx, &runtime.PodSandboxStatusRequest{
+		PodSandboxId: podId,
+	})
+
+	if status, ok := status.FromError(err); !ok || status.Code() != codes.NotFound {
+		t.Fatalf("PodSandboxStatus did not return expected errorStatus: %s", err)
+	}
+}
+
+// Utility function to test removal of a sandbox with one container and no previous call to stop the pod or container
+func runContainerInSandboxTest(t *testing.T, tc *TestConfig) {
+	client := newTestRuntimeClient(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	request := getRunPodSandboxRequest(t, tc.runtimeHandler)
+
+	podId := runPodSandbox(t, client, ctx, &request)
+	containerId := createContainerInSandbox(t, client, ctx, podId, tc.containerName, tc.containerImage, tc.cmd, map[string]string{}, []*runtime.Mount{}, request.Config)
+
+	startContainer(t, client, ctx, containerId)
+
+	removePodSandbox(t, client, ctx, podId)
```

Comment:
So one issue here is that if calling `startContainer` causes a panic, the sandbox would be leaked. Typically we defer `removePodSandbox` so that even in the event of a panic, the sandbox and its resources would be cleaned up properly. I know that here you're explicitly calling it because immediately after we want to check the status and ensure it has been removed. One way to get around this issue is to move out the code that starts a pod and container into a separate helper function, you can add a defer func there. That helper function could just return whatever information you need in the caller (here the pod ID and container ID, however you'd like) and be called in here instead to get the expected behavior.

Reply:
good point and suggestion 👍